joomla/joomla-cms
# [5.2] SEF: Enforce suffix

Pull Request for Issue #15025 .

### Summary of Changes
Joomla has been improving its SEO performance constantly and one issue which is still open is the behavior of the suffix. Right now, you can access a URL in Joomla with or without suffix when the option is enabled in the global configuration. This PR introduces a new setting the SEF system plugin which enforces a consistent suffix behavior.

When SEF URLs are enabled, the suffix is enabled and this option is enabled, Joomla will always redirect GET requests to a URL with a suffix if none is present. It will also redirect URLs with a query `format` parameter to the nicer URL and replace the "nice" suffix with the `format` parameter if the two collide.

With Joomla 6.0 the option to switch this on/off should be removed again and this should be the default behavior, which would then be added to `SiteRouter::parseFormat()`. Right now this is YASO (Yet Another Stupid Option) to allow people to test this in live systems and to switch it off if we encounter unforseen issues. The time from 5.1 to 6.0 could be seen as a grace period.

~~This PR depends on #42692.~~

I'd like to thank [djumla GmbH](https://www.djumla.de) for sponsoring this feature.


### Testing Instructions
1. Make sure that you have PR #42692 in your test installation
2. Apply the patch
3. Enable SEF URLs and the "Add Suffix to URL" options in global configuration
4. Go to "System - SEF" plugin and enable the option to "Enforce Suffix"
5. Open a URL in the frontend and see that URLs ending in a `/` don't have a suffix, everything else has a suffix.
6. Take a URL with a suffix and remove the suffix. See that Joomla redirects back to a URL with `.html`.
7. Take a URL with a suffix and add `?format=feed` to it, for example for a category blog view. See that you get the feed output with the right URL.


### Link to documentations
Please select:
- [X] Documentation link for docs.joomla.org: https://docs.joomla.org/Search_Engine_Friendly_URLs
